internal/mongodoc/doc.go

```diff
@@ -181,8 +194,20 @@ type BaseEntity struct {
 	CommonInfo map[string][]byte `bson:",omitempty" json:",omitempty"`
 }
 
-func (e *BaseEntity) CanonicalURL() *charm.URL {
-	return e.URL
+// CanonicalURLs returns the canonical URLs that can be used
+// to look up the entity. For non-promulgated entities,
+// there will only be one element in the result.
+func (e *BaseEntity) CanonicalURLs() []*charm.URL {
+	// Larger capacity than strictly necessary to allow
+	// a caller to append to the returned slice without allocating.
```

Comment:
This is really nasty. I expect you know of a situation where you expect that to be the case, but It seems nasty to put any kind of knowledge about the calling context into the function. If you must do this then please make minimum capacity a parameter.